pkg/infrastructure/aws/vpc.go

```diff
@@ -0,0 +1,1152 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to create private subnets: %w", err)
+		}
+
+		privateRouteTables = make([]*ec2.RouteTable, 0, len(vpcInput.zones))
+		for _, zone := range vpcInput.zones {
+			var natGwID *string
+			if len(publicSubnetZoneMap) > 0 {
+				natGw, err := state.ensureNatGateway(ctx, logger, ec2Client, publicSubnetZoneMap[zone], zone)
+				if err != nil {
+					return nil, fmt.Errorf("failed to create NAT gateway for zone (%s): %w", zone, err)
+				}
+				natGwID = natGw.NatGatewayId
+			}
+
+			privateRouteTable, err := state.ensurePrivateRouteTable(ctx, logger, ec2Client, natGwID, privateSubnetZoneMap[zone], zone)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create private route table for zone (%s): %w", zone, err)
+			}
+			endpointRouteTableIDs = append(endpointRouteTableIDs, privateRouteTable.RouteTableId)
+			privateRouteTables = append(privateRouteTables, privateRouteTable)
+		}
+	} else {
+		privateSubnetIDs = aws.StringSlice(vpcInput.privateSubnetIDs)
+		privateSubnetZoneMap = make(map[string]*string, len(privateSubnetIDs))
+		result, err := ec2Client.DescribeSubnetsWithContext(ctx, &ec2.DescribeSubnetsInput{
+			SubnetIds: privateSubnetIDs,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve user subnets: %w", err)
+		}
+		for _, subnet := range result.Subnets {
+			privateSubnetZoneMap[aws.StringValue(subnet.AvailabilityZone)] = subnet.SubnetId
+		}
+	}
+
+	if len(vpcInput.edgeZones) > 0 {
+		_, subnetZoneMap, err := state.ensureEdgePrivateSubnets(ctx, logger, ec2Client, networks.edge.private)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, zone := range vpcInput.edgeZones {
+			// Lookup the index of the parent zone from a given local zone
+			// name, getting the index for the route table id for that zone
+			// (parent). When not found (parent zone's gateway does not exist),
+			// the first route table will be used.
+			// Example: edgeParentMap = {"us-east-1-nyc-1a": 0}
+			tableIdx, found := vpcInput.edgeParentMap[zone]
+			if !found {
+				tableIdx = 0
+			}
+			table := privateRouteTables[tableIdx]
+			subnetID := subnetZoneMap[zone]
+			createdOrFoundMsg := "Edge subnet already associated with route table"
+			if !hasAssociatedSubnet(table, subnetID) {
+				createdOrFoundMsg = "Associated edge subnet with route table"
+				_, err := ec2Client.AssociateRouteTableWithContext(ctx, &ec2.AssociateRouteTableInput{
+					RouteTableId: table.RouteTableId,
+					SubnetId:     subnetID,
+				})
+				if err != nil {
+					return nil, fmt.Errorf("failed to associate edge subnet (%s) to route table (%s): %w", aws.StringValue(subnetID), aws.StringValue(table.RouteTableId), err)
+				}
+			}
+			logger.WithFields(logrus.Fields{
+				"subnet":      aws.StringValue(subnetID),
+				"route table": aws.StringValue(table.RouteTableId),
+			}).Infoln(createdOrFoundMsg)
```

Comment:
Ops, this was supposed to go together with https://github.com/openshift/installer/pull/7676/commits/afcb170c6209ed09d066f3065bd89ba34437cc32. I'll fix once I have other feedback to work on.